### ISSUE TITLE
Flickering/Frozen windows with offscreen decoding

### ISSUE THREAD

**Issue by ToppDev** [Xpra-org/xpra-html5#212]
### Summary

Since updating **xpra** from `4.3.4-r0-1` to `4.4.3-r0-1` and **xpra-html5** from `5.0-r1276-1` to `6.1-r5-1` windows flicker in all browsers except Chrome. Additionally, windows freeze in all browsers after some time (also in Chrome). See screenshots below.

- When using the xpra client to connect, there is not flickering/freezes. So it is related to the browser.
- When disabling `offscreen decoding` the flickering is gone and also no freezes.

### Steps to reproduce

1. Connect to the xpra server in a browser e.g. Firefox
2. Start any application there, e.g. a terminal
3. Click or type in the window to make it flicker

### Screenshots

<details>
<summary>Flickering window on user input (in Firefox)</summary>

![Peek 2022-12-14 13-44](https://user-images.githubusercontent.com/17766093/207599188-a108e677-e57e-4243-8909-da5fe59e7f2e.gif)

</details>

<details>
<summary>Frozen window (in Chrome)</summary>

![Peek 2022-12-14 14-17](https://user-images.githubusercontent.com/17766093/207605984-9753845a-ed69-451b-8484-9dc74b400622.gif)

</details>

### Relevant logs
<details>
<summary>xpra showconfig (only non-default values)</summary>

```
bandwidth-limit       (used)   = '0'                               <class 'str'>
bandwidth-limit      (default) = 'auto'                            <class 'str'>

compressors           (used)   = 'lz4', 'zlib', 'brotli'           <class 'list'>
compressors          (default) = 'none', 'lz4', 'zlib', 'brotli'   <class 'list'>

exit-with-client      (used)   = True                              <class 'bool'>
exit-with-client     (default) = False                             <class 'bool'>

file-transfer         (used)   = 'no'                              <class 'str'>
file-transfer        (default) = 'auto'                            <class 'str'>

html                  (used)   = 'no'                              <class 'str'>
html                 (default) = 'auto'                            <class 'str'>

mdns                  (used)   = False                             <class 'bool'>
mdns                 (default) = True                              <class 'bool'>

open-files            (used)   = 'no'                              <class 'str'>
open-files           (default) = 'auto'                            <class 'str'>
open-url              (used)   = 'no'                              <class 'str'>
open-url             (default) = 'auto'                            <class 'str'>

pdf-printer           (used)   = ''                                <class 'str'>
pdf-printer          (default) = '/usr/share/ppd/cupsfilters/Generic-PDF_Printer-PDF.ppd'  <class 'str'>

postscript-printer    (used)   = ''                                <class 'str'>
postscript-printer   (default) = 'drv:///sample.drv/generic.ppd'   <class 'str'>
printing              (used)   = 'no'                              <class 'str'>
printing             (default) = 'yes'                             <class 'str'>

socket-dirs           (used)   = '$XDG_RUNTIME_DIR/xpra', '/run/xpra', '~/.xpra'  <class 'list'>
socket-dirs          (default) = '/run/user/0/xpra', '/run/xpra', '~/.xpra'  <class 'list'>

sync-xvfb             (used)   = False                             <class 'bool'>
sync-xvfb            (default) =                                   <class 'NoneType'>

webcam                (used)   = 'no'                              <class 'str'>
webcam               (default) = 'auto'                            <class 'str'>

xvfb                  (used)   = 'Xvfb +extension GLX +extension Composite -screen 0  8192x4096x24+32 -nolisten tcp -noreset -auth $XAUTHORITY  -dpi 96'  <class 'str'>
xvfb                 (default) = 'Xvfb +extension GLX +extension Composite -screen 0 8192x4096x24+32 -nolisten tcp -noreset -auth $XAUTHORITY -dpi 96'  <class 'str'>
```

</details>

<details>
<summary>/etc/defaults/xpra</summary>

```
BIND_TCP=0.0.0.0:10000
#BIND_TCP=0.0.0.0:80
TCP_AUTH=sqlite:filename=/etc/xpra/xpra-auth.sdb
BIND=none
AUTH=none
HTML=on
#DEBUG=auth,proxy,util,x11
#DEBUG=auth,proxy,x11
DEBUG=auth
#skip datetime prefix since the log goes to a system logger:
XPRA_LOG_FORMAT="%(message)s"

#permissions and group ownership of /run/xpra
XPRA_SOCKET_DIR_MODE=775
XPRA_SOCKET_DIR_GROUP=xpra
```

</details>

<details>
<summary>/usr/lib/systemd/system/xpra.service</summary>

```
[Unit]
Description=Xpra System Server
Wants=avahi-daemon.socket
Documentation=https://github.com/Xpra-org/xpra/blob/master/docs/Usage/Service.md man:xpra
After=network.target xpra.socket
Requires=xpra.socket

[Service]
Type=simple
EnvironmentFile=-/etc/default/xpra
ExecStart=/usr/bin/xpra proxy :14500 --daemon=no \
    --bind-tcp=${BIND_TCP} --tcp-auth=${TCP_AUTH} \
    --socket-dirs=/run/xpra --socket-permissions=666 \
    --log-dir=/var/log --pidfile=/run/xpra/proxy.pid --debug=${DEBUG} \
    --html=${HTML}
SuccessExitStatus=0 143
Restart=on-abnormal
PIDFile=/run/xpra/proxy.pid
ProtectSystem=strict
ReadWritePaths=/run/xpra /tmp
#PrivateDevices=true
ProtectKernelTunables=true
ProtectControlGroups=true

[Install]
WantedBy=multi-user.target
```

</details>

### Environment info

- OS: Debian GNU/Linux 11 (bullseye) 5.15.74-1 x86_64
- xpra: 4.4.3-r0-1
- xpra-html5: 6.1-r5-1
- Firefox: Mozilla Firefox 107.0.1
- Chrome: Google Chrome 108.0.5359.94

**Comment by totaam** [comment]
@TijZwa have you seen this?

**Comment by TijZwa** [comment]
@ToppDev Can you also post a screenshot of the developer console in Chrome? Does it report any errors?

**Comment by TijZwa** [comment]
Misread the issue; it happens when using browsers *other*  then Chrome. I can reproduce.

**Comment by ToppDev** [comment]
Yes it is 2 things happening
- flickering in all browsers **except** Chrome
- unresponsive/freezing windows in **all** browsers

but both things are solved by disabling `offscreen decoding`

## Chrome Console log

Here a video and chrome console log for the freezing behavior.

https://user-images.githubusercontent.com/17766093/207822651-2ef774c1-89ff-4b91-87a9-7d9859a96e54.mp4

<details>
<summary>Console output</summary>

- Open Thunar window
    ```javascript
    OffscreenDecodeWorker.js:246 canvas transfer for window 18 : OffscreenCanvas {width: 1042, height: 482, oncontextlost: null, oncontextrestored: null} false
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 18 32 419 1042 482 {has-alpha: true, title: 'mitarbeiter', class-instance: Array(2), window-type: Array(1), size-constraints: {…}, …} false false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    PaintWorker.js:188 canvas transfer for window 18 : OffscreenCanvas {width: 1042, height: 482, oncontextlost: null, oncontextrestored: null} false
    ```
- Go with mouse over the items in the right list
    ```javascript
    Utilities.js:1 decode error on  throttle packet sequence 42 : TypeError: Failed to execute 'postMessage' on 'Worker': Failed to convert value to 'object'.
    Utilities.js:1 decode error packet: (7) ['damage-sequence', 42, 18, 171, 121, -1, "TypeError: Failed to execute 'postMessage' on 'Worker': Failed to convert value to 'object'."]
    ```
- Click on the items in the right list (no additional log)
- Move mouse to the left list, to the item `Desktop`. A tooltip is appearing.
    ```javascript
    OffscreenDecodeWorker.js:246 canvas transfer for window 19 : OffscreenCanvas {width: 185, height: 45, oncontextlost: null, oncontextrestored: null} false
    PaintWorker.js:188 canvas transfer for window 19 : OffscreenCanvas {width: 185, height: 45, oncontextlost: null, oncontextrestored: null} false
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 19 21 602 185 45 {title: 'Thunar', class-instance: Array(2), transient-for-xid: 10485767, transient-for: 18, window-type: Array(1), …} true false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    ```
- Click on the item `Desktop`
    ```javascript
    Utilities.js:1 lost window, was tray= false
    Utilities.js:1 lost window 19 , remaining:  ['18']
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 18 32 419 1042 482 {has-alpha: true, title: 'mitarbeiter', class-instance: Array(2), window-type: Array(1), size-constraints: {…}, …} false false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    OffscreenDecodeWorker.js:246 canvas transfer for window 20 : OffscreenCanvas {width: 185, height: 45, oncontextlost: null, oncontextrestored: null} false
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 20 16 593 185 45 {title: 'Thunar', class-instance: Array(2), transient-for-xid: 10485767, transient-for: 18, window-type: Array(1), …} true false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    PaintWorker.js:188 canvas transfer for window 20 : OffscreenCanvas {width: 185, height: 45, oncontextlost: null, oncontextrestored: null} false
    ```
- Move mouse to the item `Trash`. A tooltip is appearing.
    ```javascript
    Utilities.js:1 lost window, was tray= false
    Utilities.js:1 lost window 20 , remaining:  ['18']
    OffscreenDecodeWorker.js:246 canvas transfer for window 21 : OffscreenCanvas {width: 123, height: 45, oncontextlost: null, oncontextrestored: null} false
    PaintWorker.js:188 canvas transfer for window 21 : OffscreenCanvas {width: 123, height: 45, oncontextlost: null, oncontextrestored: null} false
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 21 13 613 123 45 {title: 'Thunar', class-instance: Array(2), transient-for-xid: 10485767, transient-for: 18, window-type: Array(1), …} true false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    ```
- Click on the item `Trash`
    ```javascript
    Utilities.js:1 lost window, was tray= false
    Utilities.js:1 lost window 21 , remaining:  ['18']
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 18 32 419 1042 482 {has-alpha: true, title: 'mitarbeiter', class-instance: Array(2), window-type: Array(1), size-constraints: {…}, …} false false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    OffscreenDecodeWorker.js:246 canvas transfer for window 22 : OffscreenCanvas {width: 123, height: 45, oncontextlost: null, oncontextrestored: null} false
    Utilities.js:1 XpraClient {container: div#screen, protocol: XpraProtocolWorkerHost, host: '<hidden-URL>', port: '443', ssl: true, …} 22 45 620 123 45 {title: 'Thunar', class-instance: Array(2), transient-for-xid: 10485767, transient-for: 18, window-type: Array(1), …} true false {} (window) => this.send_configure_window(window, {}, false) (event, window) => this.on_mousemove(event, window) (event, window) => this.on_mousedown(event, window) (event, window) => this.on_mouseup(event, window) (event, window) => this.on_mousescroll(event, window) (window) => this._window_set_focus(window) (window) => this.send([PACKET_TYPES.close_window, window.wid]) 1
    PaintWorker.js:188 canvas transfer for window 22 : OffscreenCanvas {width: 123, height: 45, oncontextlost: null, oncontextrestored: null} false
    ```


</details>

**Comment by TijZwa** [comment]
Thanks! 
The issue while freezing is a throttle package not being handled the way it should.
Throttle packages are queued when the VideoDecoder pipeline is not keeping up. 
VideoDecoder only works with chrome and secure context (localhost or SSL). 
Let me fix that.

As for the firefox issue: I can reproduce the issue.

**Comment by TijZwa** [comment]
@ToppDev Does PR #213 fixes the chrome issue?

**Comment by ToppDev** [comment]
Yes, the fix for chrome is working. The windows stay responsive

**Comment by TijZwa** [comment]
Cool.
@totaam: Please check out the video. This filemanger is sending videoframes (as the throttle package only comes from VideoDecoder)

As for Firefox; the whole offscreen pipeline is changed since version 105. (https://www.mozilla.org/en-US/firefox/105.0/releasenotes/)

**Comment by totaam** [comment]
_Support for the Offscreen Canvas DOM API with full context and font support. The [OffscreenCanvas API](https://developer.mozilla.org/docs/Web/API/OffscreenCanvas) provides a canvas that can be rendered off-screen in both Window and Web Worker contexts._
I'm on Fedora 37 which has Firefox 107.
@TijZwa does this mean that we need to disable offscreen decoding on Firefox? Or just for some versions?

**Comment by TijZwa** [comment]
> @TijZwa does this mean that we need to disable offscreen decoding on Firefox? Or just for some versions?

Yes, that is the safest bet for now. I did this in PR #215 
@ToppDev Firefox should not flicker anymore, but the performance might be worse than Chrome.

@totaam I think we should focus on moving the paint logic to WebGL instead of fixing this issue for Firefox.

**Comment by ToppDev** [comment]
I tried to test it, Firefox does not let me edit the files in the Web Developer Tools though. So I tried to manually edit the files on my server under e.g. `/usr/share/xpra/www/js/OffscreenDecodeWorker.js` and restarting the xpra and nginx service but my browser still gets the old file versions somehow. (probably the worst approach to test it, but I am not really a web developer)

However, thanks for fixing. I will just wait till the fixes make it into the next release.
